core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
> This subsumes all manner of effects as well, because a metadata key need not produce Data, it could produce M[Data].

Yeah, I'm eager to explore this with the write side of the story. Even if we don't have any means of communication metadata capabilities.